APICLIENT-3: Improve bundling size of legacy package and tidy project

maxammann opened the pull request.

maxammann committed:
APICLIENT-3: Add build script

maxammann committed:
APICLIENT-3: Switch to lodash and upgrade packages

maxammann committed:
APICLIENT-3: Add flow types

maxammann committed:
APICLIENT-3: Fix tests by fixing babel versions

maxammann committed:
APICLIENT-3: Fix linting by disable object-curly-spacing

michael-markl commented on package.json:
```diff
@@ -39,59 +39,59 @@
-    "@babel/plugin-syntax-import-meta": "^7.0.0",
-    "@babel/plugin-transform-runtime": "^7.1.0",
-    "@babel/preset-env": "^7.1.5",
+    "@babel/cli": "^7.1.2",
```
Why did you need to downgrade babel?

maxammann replied:
I tried a bit around and tried to find a version of babel which works with jest. So yeah I think the actual used version is a later version (see yarn.lock)

michael-markl commented on package.json:
```diff
@@ -39,59 +39,59 @@
     "sanitize-html-react": "^1.13.0",
-    "webpack": "^4.25.1",
-    "webpack-cli": "^3.1.2"
+    "webpack-node-externals": "^1.7.2"
```
this is a dev-dependency, isn't it?

michael-markl commented on package.json:
```diff
@@ -39,59 +39,59 @@
-    "webpack": "^4.25.1",
-    "webpack-cli": "^3.1.2"
+    "webpack-node-externals": "^1.7.2"
   }
```
I think these dependencies should have less strict version-numbers.
```
"entities": "1.x.x",
[...]
```
should be enough I guess (and better for dependents on this lib)

maxammann replied:
` ^1.2.3 will match any 1.x.x release including 1.3.0, but will hold off on 2.0.0.`

michael-markl replied:
No, it will only allow 1.2.3 upwards on minor releases. It won't allow versions 1.2.2 or below.

maxammann replied:
Yeah in this case we will get any new versions from 1.1.2 until 1.9999....9999...
Which "matching symbol" would you prefer in this case any why?

maxammann replied:
Not sure what to change it to :P

michael-markl replied:
I'd prefer 1.x.x so libs don't (unnecessarily) force dependents to use new minor versions...

maxammann replied:
To what exactly do you want to change that line? 
`"entities": "1.x.x",` is not valid afaik

michael-markl replied:
Yup it is: https://docs.npmjs.com/misc/semver#x-ranges-12x-1x-12-

michael-markl replied:
You could also just do 
```
"entities": "1"
```

maxammann replied:
k, still dont get it why we need it in this case. In order to use 1.x.x I actually have to go though every version and check if that version works with out project.

Else we won't get warnings that we a version conflict.

maxammann replied:
Actuelly I would have to do this for the other libs too... technically

michael-markl replied:
Well it's only a suggestion^^ we dont *need* to it like this.

*TLDR*: New suggestion: Use `^1.2`.

My reasoning for it is:
Once a dependent (of this lib) needs "entities", but in version "1.1.1", because there was some misbehavior introduced in "1.1.2", then this would collide with our lib if we set entities to "^1.1.2".
~The downside is, as you said, that we haven't tested our lib against all minor version of below "1.1.x". On the other hand with "^1.2.3" you also assume that any upcoming minor version will work with our lib. But then we can also trust these creators to already have had a stable API throughout "1.0.0" until "1.2.3".~

According to semver, new minor versions are backwards compatible while adding new functionalities and new patch versions are only used for bug-fixes (and new major versions for breaking changes of course).

I changed my mind to the following: If we don't want to test our lib to work with all minor versions <1.2.x, we shouldn't set the version range to "1.x.x" (because 1.2.0 could have introduced some new functionality that we depend on), but what we can do is set it to "^1.2" instead.
 So we can specify that our lib depends on the API of major version 1 and minor version 2 (which allows any minor version >= 2 according to semver)  --- while ignoring the actual patch version: that's up to the dependent of our lib whether he uses the latest patches of "entities" or not

michael-markl commented on package.json:
```diff
@@ -39,59 +39,59 @@
     "is-url": "^1.2.4",
-    "lodash-es": "^4.17.11",
-    "moment": "^2.22.2",
+    "lodash": "^4.17.11",
```
I don't know about the current output size, but if its too large, you could consider using `babel-plugin-lodash` to try to minimize it further

maxammann replied:
For now it is oke :)

maxammann committed:
APICLIENT-3: Fix babel configs to use es6 modules except for tests

maxammann committed:
APICLIENT-3: Move node-externals to dev dependencies

maxammann commented:
@klinzo ping

steffenkleinle commented on src/.eslintrc:
```diff
@@ -0,0 +1,15 @@
+    "@integreat-app/eslint-config-babel"
+  ],
+  "rules": {
+    "object-curly-spacing": ["off", "never"]
```
Shouldn't this rule rather be added to `eslint-config-flow`?

maxammann replied:
See here: https://issues.integreat-app.de/browse/LINT-10
This needs to be fixed yes.

maxammann committed:
APICLIENT-3: Change version ranges as suggested by Michael Markl
My reasoning for it is:
Once a dependent (of this lib) needs "entities", but in version "1.1.1",
because there was some misbehavior introduced in "1.1.2", then this
would collide with our lib if we set entities to "^1.1.2".

According to semver, new minor versions are backwards compatible while
adding new functionalities and new patch versions are only used for
bug-fixes (and new major versions for breaking changes of course).

I changed my mind to the following: If we don't want to test our lib to
work with all minor versions <1.2.x, we shouldn't set the version range
to "1.x.x" (because 1.2.0 could have introduced some new functionality
that we depend on), but what we can do is set it to "^1.2" instead.
So we can specify that our lib depends on the API of major version 1 and
minor version 2 (which allows any minor version >= 2 according to
semver) --- while ignoring the actual patch version: that's up to the
dependent of our lib whether he uses the latest patches of "entities"
or not
